lib/src/types/text_search.dart

```diff
@@ -116,4 +118,255 @@ class TsVectorType extends Type<TsVector> {
   }
 }
 
-class TsQuery {}
+/// The `tsquery` type represents a search query for text search.
+///
+/// https://www.postgresql.org/docs/current/datatype-textsearch.html
+/// https://www.postgresql.org/docs/current/textsearch.html
+abstract class TsQuery {
+  int get _itemCount;
+  void _write(PgByteDataWriter writer);
+
+  static TsQuery lexeme(String text) => _LexemeTsQuery._(text, 0, 0);
+  static TsQuery not(TsQuery query) => _NotTsQuery(query);
+
+  TsQuery or(TsQuery other) => _OrTsQuery._()
+    .._add(this)
+    .._add(other);
+
+  TsQuery and(TsQuery other) => _AndTsQuery._()
+    .._add(this)
+    .._add(other);
+
+  TsQuery followedBy(TsQuery other, {int? distance}) =>
+      _PhraseTsQuery(this, other, distance ?? 1);
+}
+
+class TsQueryType extends Type<TsQuery> {
+  const TsQueryType() : super(TypeOid.tsquery);
+
+  EncodeOutput encode(EncodeInput input) {
+    final v = input.value as TsQuery;
+    final writer = PgByteDataWriter(encoding: input.encoding);
+    writer.writeUint32(v._itemCount);
+    v._write(writer);
+
+    final bytes = writer.toBytes();
+    return EncodeOutput.bytes(bytes);
+  }
+
+  TsQuery? decode(DecodeInput input) {
```

Comment:
Does this ever return null? It just seems to throw for invalid bytes

Reply:
I am not yet sure what to return if the first item count is zero. Maybe that is a valid byte sequence indicating an empty query...